File: examples/vbnet/TabletopWeatherStation.vb
Imports System
Imports Tinkerforge

Module WeatherStation
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223

    Private ipcon As IPConnection = Nothing
    Private brickletLCD As BrickletLCD128x64 = Nothing
    Private brickletAirQuality As BrickletAirQuality = Nothing

    Sub AllValuesCB(ByVal sender As BrickletAirQuality, ByVal iaqIndex As Integer, _
                    ByVal iaqIndexAccuracy As Byte, ByVal temperature As Integer, _
                    ByVal humidity As Integer, ByVal airPressure As Integer)
        If brickletLCD IsNot Nothing Then
			brickletLCD.WriteLine(2, 0, String.Format("IAQ:      {0,6:###}",    iaqIndex))
			' 0xF8 == ° on LCD 128x64 charset
			brickletLCD.WriteLine(3, 0, String.Format("Temp:     {0,6:##.00} {1}C", temperature/100.0, Chr(&HF8)))
			brickletLCD.WriteLine(4, 0, String.Format("Humidity: {0,6:##.00} %RH",  humidity/100.0))
			brickletLCD.WriteLine(5, 0, String.Format("Air Pres: {0,6:####.0} mbar", airPressure/100.0))
        End If
    End Sub

    Sub EnumerateCB(ByVal sender As IPConnection, ByVal uid As String, _
                    ByVal connectedUid As String, ByVal position As Char, _
                    ByVal hardwareVersion() As Short, ByVal firmwareVersion() As Short, _
                    ByVal deviceIdentifier As Integer, ByVal enumerationType As Short)
        If enumerationType = IPConnection.ENUMERATION_TYPE_CONNECTED Or _
           enumerationType = IPConnection.ENUMERATION_TYPE_AVAILABLE Then
            If deviceIdentifier = BrickletLCD128x64.DEVICE_IDENTIFIER Then
                Try
					' Initialize newly enumerated LCD128x64 Bricklet
                    brickletLCD = New BrickletLCD128x64(UID, ipcon)
                    brickletLCD.ClearDisplay()
					brickletLCD.WriteLine(0, 0, "   Weather Station")
                    System.Console.WriteLine("LCD 128x64 initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("LCD 128x64 init failed: " + e.Message)
                    brickletLCD = Nothing
                End Try
            Else If deviceIdentifier = BrickletAirQuality.DEVICE_IDENTIFIER Then
                Try
				    ' Initialize newly enumaratedy Air Quality Bricklet and configure callbacks
                    brickletAirQuality = New BrickletAirQuality(UID, ipcon)
                    brickletAirQuality.SetAllValuesCallbackConfiguration(1000, false)
                    AddHandler brickletAirQuality.AllValuesCallback, AddressOf AllValuesCB
                    System.Console.WriteLine("Air Quality initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("Air Quality init failed: " + e.Message)
                    brickletAirQuality = Nothing
                End Try
            End If
        End If
    End Sub

    Sub ConnectedCB(ByVal sender As IPConnection, ByVal connectedReason as Short)
		' Eumerate again after auto-reconnect
        If connectedReason = IPConnection.CONNECT_REASON_AUTO_RECONNECT Then
            System.Console.WriteLine("Auto Reconnect")
            while True
                Try
                    ipcon.Enumerate()
                    Exit While
                Catch e As NotConnectedException
                    System.Console.WriteLine("Enumeration Error: " + e.Message)
                    System.Threading.Thread.Sleep(1000)
                End Try
            End While
        End If
    End Sub

    Sub Main()
        ipcon = New IPConnection() ' Create IP connection

		' Connect to brickd (retry if not possible)
        while True
            Try
                ipcon.Connect(HOST, PORT)
                Exit While
            Catch e As System.Net.Sockets.SocketException
                System.Console.WriteLine("Connection Error: " + e.Message)
                System.Threading.Thread.Sleep(1000)
            End Try
        End While

        AddHandler ipcon.EnumerateCallback, AddressOf EnumerateCB
        AddHandler ipcon.Connected, AddressOf ConnectedCB

		' Enumerate Bricks and Bricklets (retry if not possible)
        while True
            try
                ipcon.Enumerate()
                Exit While
            Catch e As NotConnectedException
                System.Console.WriteLine("Enumeration Error: " + e.Message)
                System.Threading.Thread.Sleep(1000)
            End Try
        End While

        System.Console.WriteLine("Press key to exit")
        System.Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module

File: examples/delphi/TabletopWeatherStation.pas
program TabletopWeatherStation;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, Device, BrickletLCD128x64, BrickletAirQuality;

const
  HOST = 'localhost';
  PORT = 4223;

type
  TTabletopWeatherStation = class
  private
    ipcon: TIPConnection;
    brickletLCD: TBrickletLCD128x64;
    brickletAirQuality: TBrickletAirQuality;
  public
    constructor Create;
    destructor Destroy; override;
    procedure ConnectedCB(sender: TIPConnection; const connectedReason: byte);
    procedure EnumerateCB(sender: TIPConnection; const uid: string;
                          const connectedUid: string; const position: char;
                          const hardwareVersion: TVersionNumber;
                          const firmwareVersion: TVersionNumber;
                          const deviceIdentifier: word; const enumerationType: byte);
    procedure AllValuesCB(sender: TBrickletAirQuality; const iaqIndex: longint;
                          const iaqIndexAccuracy: byte; const temperature: longint;
                          const humidity: longint; const airPressure: longint);
    procedure Execute;
  end;

var
  tws: TTabletopWeatherStation;

constructor TTabletopWeatherStation.Create;
begin
  ipcon := nil;
  brickletLCD := nil;
  brickletAirQuality := nil;
end;

destructor TTabletopWeatherStation.Destroy;
begin
  if (brickletLCD <> nil) then brickletLCD.Destroy;
  if (brickletAirQuality <> nil) then brickletAirQuality.Destroy;
  if (ipcon <> nil) then ipcon.Destroy;
  inherited Destroy;
end;

procedure TTabletopWeatherStation.ConnectedCB(sender: TIPConnection; const connectedReason: byte);
begin
  { Eumerate again after auto-reconnect }
  if (connectedReason = IPCON_CONNECT_REASON_AUTO_RECONNECT) then begin
    WriteLn('Auto Reconnect');
    while (true) do begin
      try
        ipcon.Enumerate;
        break;
      except
        on e: Exception do begin
          WriteLn('Enumeration Error: ' + e.Message);
          Sleep(1000);
        end;
      end;
    end;
  end;
end;

procedure TTabletopWeatherStation.EnumerateCB(sender: TIPConnection; const uid: string;
                                              const connectedUid: string; const position: char;
                                              const hardwareVersion: TVersionNumber;
                                              const firmwareVersion: TVersionNumber;
                                              const deviceIdentifier: word; const enumerationType: byte);
begin
  if ((enumerationType = IPCON_ENUMERATION_TYPE_CONNECTED) or
      (enumerationType = IPCON_ENUMERATION_TYPE_AVAILABLE)) then begin
    if (deviceIdentifier = BRICKLET_LCD_128X64_DEVICE_IDENTIFIER) then begin
      try
        { Initialize newly enumerated LCD128x64 Bricklet }
        brickletLCD := TBrickletLCD128x64.Create(UID, ipcon);
        brickletLCD.ClearDisplay;
		brickletLCD.WriteLine(0, 0, '   Weather Station');
        WriteLn('LCD 128x64 initialized');
      except
        on e: Exception do begin
          WriteLn('LCD 128x64 init failed: ' + e.Message);
          brickletLCD := nil;
        end;
      end;
    end
    else if (deviceIdentifier = BRICKLET_AIR_QUALITY_DEVICE_IDENTIFIER) then begin
      try
	    { Initialize newly enumaratedy Air Quality Bricklet and configure callbacks }
        brickletAirQuality := TBrickletAirQuality.Create(uid, ipcon);
        brickletAirQuality.SetAllValuesCallbackConfiguration(1000, true);
        brickletAirQuality.OnAllValues := {$ifdef FPC}@{$endif}AllValuesCB;
        WriteLn('Air Quality initialized');
      except
        on e: Exception do begin
          WriteLn('Air Quality init failed: ' + e.Message);
          brickletAirQuality := nil;
        end;
      end;
    end
  end;
end;

procedure TTabletopWeatherStation.AllValuesCB(sender: TBrickletAirQuality; const iaqIndex: longint;
                                              const iaqIndexAccuracy: byte; const temperature: longint;
                                              const humidity: longint; const airPressure: longint);
var text: string;
begin
  if (brickletLCD <> nil) then begin
    text := Format('IAQ:      %6d', [iaqIndex]);
    brickletLCD.WriteLine(2, 0, text);

	{ $F8 == ° on LCD 128x64 charset }
    text := Format('Temp:     %6.2f %sC', [temperature/100.0, '' + char($F8)]);
    brickletLCD.WriteLine(3, 0, text);
    text := Format('Humidity: %6.2f %%RH', [humidity/100.0]);
    brickletLCD.WriteLine(4, 0, text);
    text := Format('Air Pres: %6.2f hPa', [airPressure/100.0]);
    brickletLCD.WriteLine(5, 0, text);
  end;
end;

procedure TTabletopWeatherStation.Execute;
begin
  ipcon := TIPConnection.Create; { Create IP connection }

  { Connect to brickd (retry if not possible) }
  while (true) do begin
    try
      ipcon.Connect(HOST, PORT);
      break;
    except
      on e: Exception do begin
        WriteLn('Connection Error: ' + e.Message);
        Sleep(1000);
      end;
    end;
  end;
  ipcon.OnEnumerate := {$ifdef FPC}@{$endif}EnumerateCB;
  ipcon.OnConnected := {$ifdef FPC}@{$endif}ConnectedCB;

  { Enumerate Bricks and Bricklets (retry if not possible) }
  while (true) do begin
    try
      ipcon.Enumerate;
      break;
    except
      on e: Exception do begin
        WriteLn('Enumeration Error: ' + e.Message);
        Sleep(1000);
      end;
    end;
  end;
  WriteLn('Press key to exit');
  ReadLn;
end;

begin
  tws := TTabletopWeatherStation.Create;
  tws.Execute;
  tws.Destroy;
end.
